src/cmd/ksh93/edit/history.c

```diff
@@ -455,6 +455,10 @@ static History_t *hist_trim(History_t *hp, int n) {
             if (newp <= oldp) break;
         }
         if (!(buff = (char *)sfreserve(hist_old->histfp, SF_UNBOUND, 0))) break;
```

Comment:
`SF_UNBOUND` is the magic value -1. I admit I have no idea what `sfreserve()` is actually doing but since it's returning a pointer to the buffer what happens if you just tell it to reserve a buffer exactly the size we need? That is,

```
if (!(buff = (char *)sfreserve(hist_old->histfp, sfvalue(hist_old->histfp) + 1, 0))) break;
```

Note that there's no need to multiply by `sizeof(char)` since that is by definition one.

I suspect the bug is that the subsequent stuffing of null into the end of the buffer is violating the SFIO API and that there is probably a SFIO function that should be used to append a char (or string) to a stream that has been set to size `SF_UNBOUND` so that SFIO takes care of allocating the additional space. Which is why I want to see SFIO replaced with stdio. Who among us knows anything about SFIO?

Reply:
@krader1961 
> Note that there's no need to multiply by sizeof(char) since that is by definition one.

For historical information:
There was a time around about 1984, when AT&T UNIX (I think it was either AT&T System V Release 1 or 2) was ported to the Cray computer of that time. The C language compiler was a cheap port (meaning very simply done), and since the Cray did not have byte addressability, the compiler assigned each 'char' to its own addressable memory word on the Cray -- which at the time was 64-bits. So the sizeof(char) was 8! Hopefully that particular compiler port did not last long. I learned this at a talk given on the various ports of AT&T UNIX around then.  As an aside, AT&T UNIX SysV was (or had been) adopted by a number of other manufactures then, including HP and Amdahl.

Reply:
@DavidMorano, I'll bet you a paycheck I've used more idiosyncratic systems than anyone working on this project other than its founders. When I was a college student I used a CDC Cyber 760 with a 60 bit integer that used "truncated ASCII" (6-bit chars so no no lowercase chars and ten chars to an integer word). Which also used ones complement so it had both positive  and negative zero. I used a Harris H500 that had a three byte (24 bit) integer.

While I can appreciate the necessity of supporting such unusual platforms 20+ years ago that is no longer true. Those systems are no longer being sold. Which means we are no longer encumbered by them and can therefore simplify our code.

Reply:
@krader1961 
> I'll bet you a paycheck I've used more idiosyncratic systems than anyone working on this project other than its founders.

:-) :-) !! :-)

> ... CDC Cyber 760 ...

I think you made a typo and meant to say the "CDC Cyber 7600".

I worked on the CDC Cyber 6600 (the predecessor machine), almost identical w/ the 7600, w/ 60 bit words, 10 6-bit chars to a word, one's compliment! :-)  I was a great assembly language programmer for that machine! :-)

I also used the DEC PDP-8 with its 12-bit words! :-) :-)

We are -- sadly or not -- really showing our ages here on this forum! :-)

Reply:
@krader1961 
> Although maybe close, there is still no way I would take your bet! :-)

Reply:
@krader1961 According to sfio man page (`src/lib/libast/man/sfio.3`):

```
After a sfreserve() call, whether or not it succeeds, sfvalue(f) gives the size of the available data block.
```
So it's not possible to call `sfavlue()` before `sfreserve()`.

Reply:
Okay, but the man page also says

> n < 0: sfreserve() attempts to get at least |n| bytes into the buffer.

And since `SF_UNBOUND` is -1 that means it is reserving one byte. What I suspect is happening is it returns a buffer whose size is one only if the buffer was currently zero length otherwise it returns a pointer to the existing buffer. And the person who wrote this code expected it to reserve one __additional__ byte to make room for the nul char. The only reason it's appending that nul is so that it can do

```c
        while (*cp++) {
            ;  // empty loop
        }
```

even if the old hist buffer does not contain a nul terminated string. That block of code should be rewritten so that it doesn't need to append the nul char. There isn't any reason we should pay the cost of an expensive malloc and buffer copy just to retain the existing, bad, logic for copying an existing history buffer from the old hist file to the new one.

Also, looking at the definition of `SF_UNBOUND` I don't understand why they're using that rather than just passing `-1`. Also, from reading the man page it looks like that should just be `sfreserve(f, 0, -1)` to get a pointer to the existing buffer that we can then call `sfvalue()` to get the length of it.

Finally, grep shows a lot of places that call `sfreserve()` with `SF_UNBOUND` and the first couple I looked at seem dubious. The man page implies that symbol should only be used with functions like `sfsetbuf()` and `sftmp()` to create an unbuffered file. I'm reasonably confident that all the places passing that to `sfreserve()` should either be passing a literal `-1` or they have the same bug as this block of code.

Reply:
@krader1961 Ok, I will try to rewrite the patch.